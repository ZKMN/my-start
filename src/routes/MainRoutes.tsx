import {
  Route,
  Switch,
  RouteChildrenProps,
} from 'react-router-dom';

import { ConnectedCoreLayout } from 'containers';
import { NotFoundPage } from 'components';

import {
  routesConfig,
  getPublicRoutes,
  getPrivateRoutes,
} from './routeHelpers';

export const MainRoutes = ({
  location,
  history,
  match,
}: RouteChildrenProps) => (
  <ConnectedCoreLayout
    location={location}
    history={history}
    match={match}
  >
    <Switch>
      { getPublicRoutes(routesConfig.publicRoutes) }
      { getPrivateRoutes(routesConfig.privateRoutes) }
      <Route path="*" component={NotFoundPage} />
    </Switch>
  </ConnectedCoreLayout>
);
