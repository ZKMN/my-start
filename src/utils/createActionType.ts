export enum RequestStatuses {
  Request = "REQUEST",
  Success = "SUCCESS",
  Failure = "FAILURE",
}
export enum XHRMethod {
  Get = "GET",
  Post = "POST",
  Put = "PUT",
  Patch = "PATCH",
  Delete = "DELETE",
}

export const createActionType = (
  method: XHRMethod,
  actionName: string,
  addFailureType?: boolean,
) => {
  const actionTypes = {
    REQUEST: `${method}_${actionName}_${RequestStatuses.Request}`,
    SUCCESS: `${method}_${actionName}_${RequestStatuses.Success}`,
    ...(addFailureType && { FAILURE: `${method}_${actionName}_${RequestStatuses.Failure}` }),
  }

  return actionTypes;
};
