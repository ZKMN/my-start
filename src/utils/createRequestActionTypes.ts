export enum RequestActionStatuses {
  Request = "REQUEST",
  Success = "SUCCESS",
  Failure = "FAILURE",
}

export enum XHRMethod {
  Get = "GET",
  Post = "POST",
  Put = "PUT",
  Patch = "PATCH",
  Delete = "DELETE",
}
export interface ICreateRequestActionTypes {
  REQUEST: string;
  SUCCESS: string;
  FAILURE?: string;
}

export const createRequestActionType = (method: XHRMethod, actionName: string) => `${method}_${actionName}`;

export const createRequestActionTypes = (method: XHRMethod, actionName: string, addFailureType?: boolean) => {
  const actionTypes = {
    REQUEST: `${createRequestActionType(method, actionName)}_${RequestActionStatuses.Request}`,
    SUCCESS: `${createRequestActionType(method, actionName)}_${RequestActionStatuses.Success}`,
    ...(addFailureType && { FAILURE: `${createRequestActionType(method, actionName)}_${RequestActionStatuses.Failure}` }),
  }

  return actionTypes;
};
