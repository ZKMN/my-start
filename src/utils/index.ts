export * from "./addActionsToSagas";
export * from "./createRequestAction";
export * from "./createRequestActionTypes";
export * from "./addActionsToSagas";
export * from "./sagaUtils";
