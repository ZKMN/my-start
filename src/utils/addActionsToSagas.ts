import { XHRMethod, ICreateRequestActionTypes } from "./createRequestActionTypes";

export const getActions = [] as string[];
export const postActions = [] as string[];
export const putActions = [] as string[];
export const patchActions = [] as string[];
export const deleteActions = [] as string[];
export const commonActions = [] as string[];

export const addActionsToSagas = (actionTypes: ICreateRequestActionTypes[]) => {
  actionTypes.forEach((actionType: ICreateRequestActionTypes) => {
    if (actionType.REQUEST.includes(XHRMethod.Get)) return getActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Post)) return postActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Put)) return putActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Patch)) return patchActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Delete)) return deleteActions.push(actionType.REQUEST);
    return null;
  });
};
