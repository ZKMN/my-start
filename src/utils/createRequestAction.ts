import { ICreateRequestActionTypes } from "./createRequestActionTypes";
import { StringifiableRecord } from 'query-string';

interface TActionCallbackResult {
  type: string | undefined;
  payload?: Record<string, unknown>;
}

interface IRequestPayload {
  payload?: any; 
  routeParams?: any;
  [key: string]: unknown; 
}

type TActionCallback = (payload: any) => TActionCallbackResult;

export interface IRequestAction {
  type: string | undefined;
  payload: Record<string, unknown>;
  endpoint: string;
  responseType?: string;
  queryParams?: StringifiableRecord;
  routeParams?: Record<string, unknown>;
  successCallback: TActionCallback;
  failureCallback?: TActionCallback;
}

export const createRequestAction = (
  actionType: ICreateRequestActionTypes,
  requestEndpoint: string,
  responseType?: string,
) => (
  requestPayload?: IRequestPayload,
): IRequestAction => {

  const { payload, routeParams, ...rest } = Object(requestPayload);

  const failureCallback = (payload: any) => ({
    type: actionType.FAILURE,
    payload,
  });
  const successCallback = (payload: any) => ({
    type: actionType.SUCCESS,
    payload,
  });

  const action: IRequestAction = {
    type: actionType.REQUEST,
    queryParams: { ...rest },
    routeParams,
    payload,
    endpoint: requestEndpoint,
    responseType,
    successCallback,
    ...(actionType.FAILURE && { failureCallback }),
  };

  return action;
};
