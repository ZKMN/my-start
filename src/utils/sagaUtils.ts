import { stringifyUrl } from "query-string";
import { IRequestAction } from "utils";
// @ts-ignore
import path, { pathParams } from "path-params";

// noParams - value for parallel request if we don't need to add params to endpoint
export const addParamsToURL = (action: IRequestAction, noParams?: boolean) => {
  const { queryParams, routeParams, endpoint } = action;

  let urlWithRouteParams;

  // 2 imports and condition assignement for test, jest can't find pathParams
  // TODO: find solution to solve this problem
  if (pathParams) {
    urlWithRouteParams = pathParams(endpoint, routeParams);
  } else {
    urlWithRouteParams = path(endpoint, routeParams);
  }

  if (!noParams) {
    return stringifyUrl(
      { url: urlWithRouteParams, query: queryParams },
      { skipNull: true, skipEmptyString: true, arrayFormat: "comma" },
    );
  }

  return urlWithRouteParams;
};
