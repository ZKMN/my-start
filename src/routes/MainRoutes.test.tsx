import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { RouteChildrenProps } from 'react-router-dom';

import { MainRoutes } from './MainRoutes';
import { routerTestProps } from 'utils';

let wrapper: ShallowWrapper;

const { history, location, match } = routerTestProps();

const props: RouteChildrenProps = {
  location,
  history,
  match,
};

describe('MainRoutes', () => {
  beforeEach(() => {
    wrapper = shallow(<MainRoutes {...props} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
