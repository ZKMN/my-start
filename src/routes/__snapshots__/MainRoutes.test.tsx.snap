// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainRoutes matches snapshot 1`] = `
<CoreLayout
  history={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "1",
      "pathname": "",
      "search": "",
      "state": Object {},
    }
  }
  match={
    Object {
      "isExact": false,
      "params": Object {
        "id": "1",
      },
      "path": "/route/:id",
      "url": "/route/1",
    }
  }
>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      key="/"
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      key="/"
      path="/"
    />
    <Route
      component={[Function]}
      path="*"
    />
  </Switch>
</CoreLayout>
`;
